# CompGen
CompGen Assignments and Labs are in their designated folders. 
## Assignment 1: HMM BaseCaller and FastQC

Created a program that took an CSV file containing normalized fluorescent intensitites for each nucleotide position in each read as input and used the Viterbi algorithm to compute the most likely sequence of nucleotides for each read in FASTA format. The HMM packages in Python (numpy, HMM, pandas, matplotlib). ​Now use the CSV file under Assignment 1 (Files->Assignment Files->assignment1.csv). 

Also created a program that took a FASTQ file and printed the distribution of quality scores across all reads. You can summarize the distribution of Q scores at each base with a statistic of your choice (e.g. mean, mode, median, quantile distribution). If you’d like, you can also plot the distribution of Q scores as a box plot much like what’s generated by FASTQC. You will then run your program on the provided FASTQ file, and obtain the output from it and upload it with your submission. 

## Assignment 2: 

For this assignment, you will write code to (1) take an input genome as a FASTA file, (2) construct a suffix array/tree from it, (3) take a read as input from a FASTA file, and (4) print all positions where that read occurs in the genome using the bisection algorithm.

s = 'banana' #read the genome string

sa = sorted([s[i:] for i in range(len(s))]) #construct the suffix array library

Note what it’s doing – iterating over 1 – length of the string s; then sorting the array. This will give you the suffix array directly. But note that this is not very memory efficient if you have a large genome. So it might be easier to just directly store the sorted suffix array indices instead. I’ll let you get creative with this!

Please submit your code, and output file with all indices (locations of match) for the input files provided (you'll find this under Assignment Files - Assignment2_refgenome.fasta, Assignment2_read.fasta).

